File: backend/app/schemas/document.py
"""
Document schemas.

Pydantic models for document upload, listing, and detail responses.
"""

from datetime import datetime
from enum import StrEnum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class DocumentType(StrEnum):
    """Supported legal document types."""
    TERMS_AND_CONDITIONS = "terms_and_conditions"
    PRIVACY_POLICY = "privacy_policy"
    RENTAL_AGREEMENT = "rental_agreement"
    EMPLOYMENT_CONTRACT = "employment_contract"
    SERVICE_AGREEMENT = "service_agreement"
    LOAN_AGREEMENT = "loan_agreement"
    EULA = "eula"
    OTHER = "other"


class DocumentStatus(StrEnum):
    """Document processing status."""
    UPLOADED = "uploaded"
    EXTRACTING = "extracting"
    SEGMENTING = "segmenting"
    EMBEDDING = "embedding"
    ANALYZING = "analyzing"
    COMPLETED = "completed"
    FAILED = "failed"


class DocumentUploadResponse(BaseModel):
    """Response after a successful document upload."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    filename: str
    original_filename: str
    file_type: str
    file_size: int
    document_type: DocumentType
    status: DocumentStatus
    created_at: datetime


class DocumentListItem(BaseModel):
    """Condensed document info for list views."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    original_filename: str
    document_type: DocumentType
    status: DocumentStatus
    file_size: int
    is_scanned: bool = False
    clause_count: int = 0
    overall_risk_score: int | None = None
    created_at: datetime


class DocumentDetail(BaseModel):
    """Full document details including metadata."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    original_filename: str
    filename: str
    file_type: str
    file_size: int
    document_type: DocumentType
    status: DocumentStatus
    jurisdiction: str | None = None
    is_scanned: bool = False
    raw_text: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict, validation_alias="metadata_")
    created_at: datetime
    updated_at: datetime | None = None

File: frontend/src/types/document.ts
/**
 * TypeScript type definitions for documents.
 */

export type DocumentType =
  | "terms_and_conditions"
  | "privacy_policy"
  | "rental_agreement"
  | "employment_contract"
  | "service_agreement"
  | "loan_agreement"
  | "eula"
  | "other";

export type DocumentStatus =
  | "uploaded"
  | "extracting"
  | "segmenting"
  | "embedding"
  | "analyzing"
  | "completed"
  | "failed";

export interface Document {
  id: string;
  original_filename: string;
  filename: string;
  file_type: string;
  file_size: number;
  document_type: DocumentType;
  status: DocumentStatus;
  jurisdiction: string | null;
  is_scanned: boolean;
  raw_text: string | null;
  metadata: Record<string, unknown>;
  clause_count?: number;
  overall_risk_score?: number | null;
  created_at: string;
  updated_at: string | null;
}

export interface DocumentListItem {
  id: string;
  original_filename: string;
  document_type: DocumentType;
  status: DocumentStatus;
  file_size: number;
  is_scanned: boolean;
  clause_count: number;
  overall_risk_score: number | null;
  created_at: string;
}

export interface Clause {
  id: string;
  document_id: string;
  clause_index: number;
  clause_number: string | null;
  title: string | null;
  content: string;
  category: string | null;
  start_page: number | null;
  end_page: number | null;
  start_char: number | null;
  end_char: number | null;
  metadata: Record<string, unknown>;
}

export const DOCUMENT_TYPE_LABELS: Record<DocumentType, string> = {
  terms_and_conditions: "Terms & Conditions",
  privacy_policy: "Privacy Policy",
  rental_agreement: "Rental Agreement",
  employment_contract: "Employment Contract",
  service_agreement: "Service Agreement",
  loan_agreement: "Loan Agreement",
  eula: "EULA",
  other: "Other",
};

export const STATUS_LABELS: Record<DocumentStatus, string> = {
  uploaded: "Uploaded",
  extracting: "Extracting Text",
  segmenting: "Segmenting Clauses",
  embedding: "Generating Embeddings",
  analyzing: "Analyzing",
  completed: "Completed",
  failed: "Failed",
};
